### PR TITLE
mv from 'post' to 'media' and change data schema & replace gm to sharp & gen 4 size for livephoto

### PR DESCRIPTION
**Pull request by uniray7**
1. mv from 'post' to 'media'
2. change data schema
3. add Jenkinsfile
4. modify Dockerfile
5. replace gm to sharp 
6. gen 4 size for livephoto

### CODE REVIEW

File: src/media.js [hunk 1]
```diff
@@ -129,35 +110,36 @@ var processImageAsync = P.promisify(function(params, callback) {
...
-                             params.postId + '.jpg' ];
+    .then(function(result) { // add srcURL and srcDownUrl
+//      response = merge({}, response, {
+//          src:{
+//              srcUrl: result.location,
+//              srcDownloadUrl: result.location
+//          }
+//      });
+      var thumbImgKeyArr = [ params.shardingKey, 'media', params.mediaId, 'pano',
+                             'thumb' + '.jpg' ];
```
Comment (weitingchou) [inline src/media.js]: Can't see why it needs to do concatenation of 'thumb' and '.jpg' rather then just use 'thumb.jpg'.

File: src/media.js [hunk 2]
```diff
@@ -344,8 +344,8 @@ var deletePostImages = function(job) {
 };
 
 function addTo(worker) {
-  worker.addFunction('postProcessingPanoPhoto', postProcessingPanoPhoto, { timeout: config.defaultTimeout });
-  worker.addFunction('postProcessingLivePhoto', postProcessingLivePhoto, { timeout: config.defaultTimeout });
+  worker.addFunction('mediaProcessingPanoPhoto', mediaProcessingPanoPhoto, { timeout: config.defaultTimeout });
+  worker.addFunction('mediaProcessingLivePhoto', mediaProcessingLivePhoto, { timeout: config.defaultTimeout });
   worker.addFunction('deletePostImages', deletePostImages, { timeout: config.defaultTimeout });
```
Comment (SuJiaKuan) [inline src/media.js]: We should also rename `deletePostImages`

**Comment by SuJiaKuan** [comment]
In `src/media.js`, please add spaces before and after + operator.

File: src/media.js [hunk 3]
```diff
@@ -0,0 +1,366 @@
+var P = require('bluebird');
+var merge = require('lodash/merge');
+var moment = require('moment');
+var async = require('async');
+var urlencode = require('urlencode');
+var sharp = require('sharp');
```
Comment (SuJiaKuan) [inline src/media.js:6]: Please use `npm install --save sharp` to add dependencies in `package.json`

File: src/media.js [hunk 4]
```diff
@@ -172,8 +174,55 @@ var mediaProcessingPanoPhoto = function(job) {
 
 function processLivePhotoSrc(params, callback) {
   var response = {};
+  var sizeQualList;  
   var srcImgKeyArr = [ params.shardingKey, 'media', params.mediaId, 'live', 'src' + (params.image.hasZipped ? '.jpg.zip' : '.jpg') ];
   var srcImgBuf = new Buffer(params.image.buffer, 'base64');
+  
+  function calSizeQualityList(imgMetadata){
```
Comment (SuJiaKuan) [inline src/media.js]: Please add a space before left bracket.

File: src/media.js [hunk 5]
```diff
@@ -172,8 +174,55 @@ var mediaProcessingPanoPhoto = function(job) {
...
+    var list = [];
+    var element ;
+    var imgWidth, imgHeight;  
+    if (imgMetadata.orientation>4) // Exif Orientation Tag ref:http://sylvana.net/jpegcrop/exif_orientation.html
```
Comment (SuJiaKuan) [inline src/media.js]: 1. Please add spaces before and after `>` operator
2. Please do not change indent style

File: src/media.js [hunk 6]
```diff
@@ -172,8 +174,55 @@ var mediaProcessingPanoPhoto = function(job) {
...
+    {
+        imgWidth = imgMetadata.height;
+        imgHeight = imgMetadata.width;
+    }
+    else{
+        imgWidth = imgMetadata.width;
+        imgHeight = imgMetadata.height;
+    }
+    
+    if(imgWidth> 600){
```
Comment (SuJiaKuan) [inline src/media.js]: 1. Please add a space before `>` operator
2. Please a space before left bracket

File: src/media.js [hunk 7]
```diff
@@ -172,8 +174,55 @@ var mediaProcessingPanoPhoto = function(job) {
...
+      list.push({width:600, height:Math.round(imgHeight*600/imgWidth), quality:70});
+      list.push({width:480, height:Math.round(imgHeight*480/imgWidth), quality:75});
+      list.push({width:360, height:Math.round(imgHeight*360/imgWidth), quality:80});
+      list.push({width:240, height:Math.round(imgHeight*240/imgWidth), quality:90});
+    }
+    else if((imgWidth <= 600) && (imgWidth>480))
```
Comment (SuJiaKuan) [inline src/media.js]: Please do not change indent style

File: src/media.js [hunk 8]
```diff
@@ -183,43 +232,47 @@ function processLivePhotoSrc(params, callback) {
     }
     return P.resolve(srcImgBuf);
   })
-  .then(function(imgBuf) {
-    var parsedImgArr = Buffer(imgBuf, 'binary').toString('binary').split(params.image.arrayBoundary);
+  .then(function(srcImgBuf){
+    return new P(function(resolve, reject){
+      var parsedImgArr = Buffer(srcImgBuf, 'binary').toString('binary').split(params.image.arrayBoundary);
+      sharp(Buffer(parsedImgArr[0],'binary'))
```
Comment (SuJiaKuan) [inline src/media.js:229]: Please add a space after comma

File: src/media.js [hunk 9]
```diff
@@ -183,43 +232,47 @@ function processLivePhotoSrc(params, callback) {
...
+        .metadata(function(err, metadata){
+          if(err){return reject(err);}
```
Comment (SuJiaKuan) [inline src/media.js]: Please add spaces to keep the coding style

File: src/media.js [hunk 10]
```diff
@@ -183,43 +232,47 @@ function processLivePhotoSrc(params, callback) {
...
-          }, function(err) {
-            if (err) { return callback(err); }
-            callback();
-          });
-        });
-      }, function(err) {
+      async.forEachOf(parsedImgArr, function(image, imgIndex, asyncParsedImgCb) {
+        var imgObj = sharp(Buffer(image, 'binary'));
+        var rotatedImgObj = imgObj.rotate();
+        async.forEachOf(sizeQualList, function(sizeQual, listIndex, asyncSizeQualCb){
```
Comment (SuJiaKuan) [inline src/media.js]: Please add a space before left bracket

File: src/media.js [hunk 11]
```diff
@@ -183,43 +232,47 @@ function processLivePhotoSrc(params, callback) {
...
+          rotatedImgObj
+          .resize(sizeQual.width, sizeQual.height)
+          .quality(sizeQual.quality)
+          .toBuffer(function(err, outputBuf){
+            var keyArr = [ params.shardingKey, 'media', params.mediaId, 'live', sizeQual.width+'X'+sizeQual.height, imgIndex + '.jpg' ];
```
Comment (SuJiaKuan) [inline src/media.js]: Please add spaces before and after `+` operator

File: src/media.js [hunk 12]
```diff
@@ -183,43 +232,47 @@ function processLivePhotoSrc(params, callback) {
...
+              if (err) { return asyncSizeQualCb(err); }
+              asyncSizeQualCb();
+            }); // store.create
+          }); // .toBuffer
+        },
+        function(err){
+          if (err) { return reject(err); }
+          asyncParsedImgCb();  
+        }); // async.forEachOf
+      },function(err){
```
Comment (SuJiaKuan) [inline src/media.js]: Please add spaces to keep the coding style

File: src/media.js [hunk 13]
```diff
@@ -183,43 +232,47 @@ function processLivePhotoSrc(params, callback) {
...
-        });
-      }, function(err) {
+      async.forEachOf(parsedImgArr, function(image, imgIndex, asyncParsedImgCb) {
+        var imgObj = sharp(Buffer(image, 'binary'));
+        var rotatedImgObj = imgObj.rotate();
+        async.forEachOf(sizeQualList, function(sizeQual, listIndex, asyncSizeQualCb){
+          rotatedImgObj
+          .resize(sizeQual.width, sizeQual.height)
+          .quality(sizeQual.quality)
+          .toBuffer(function(err, outputBuf){
```
Comment (SuJiaKuan) [inline src/media.js]: Please add a space before left bracket

File: src/media.js [hunk 14]
```diff
@@ -183,43 +232,47 @@ function processLivePhotoSrc(params, callback) {
...
+        }); // async.forEachOf
+      },function(err){
         if (err) { return reject(err); }
-        resolve({ count: parsedImgArr.length,
-                  quality: [params.image.width+'X'+params.image.height, DEFAULT_LIVE_LOW_DIMENSION.width+'X'+DEFAULT_LIVE_LOW_DIMENSION.height]});
+        var imgSizeList =[];
+        for(var i=0; i<sizeQualList.length; i++)
+        {
+            imgSizeList.push(sizeQualList[i].width+'X'+sizeQualList[i].height);
+        }
```
Comment (SuJiaKuan) [inline src/media.js]: Maybe it's better to use `map` function, such as:

```
var imgSizeList = sizeQualList.map(function(sizeQual) {
  return sizeQual.width + 'X' + sizeQual.height);
});
```

File: src/media.js [hunk 15]
```diff
@@ -183,43 +232,47 @@ function processLivePhotoSrc(params, callback) {
...
-  .then(function(imgBuf) {
-    var parsedImgArr = Buffer(imgBuf, 'binary').toString('binary').split(params.image.arrayBoundary);
+  .then(function(srcImgBuf){
+    return new P(function(resolve, reject){
+      var parsedImgArr = Buffer(srcImgBuf, 'binary').toString('binary').split(params.image.arrayBoundary);
+      sharp(Buffer(parsedImgArr[0],'binary'))
+        .metadata(function(err, metadata){
+          if(err){return reject(err);}
+          sizeQualList = calSizeQualityList(metadata);
+          resolve(parsedImgArr);
```
Comment (SuJiaKuan) [inline src/media.js:233]: sizeQualList is declared too early, maybe we can modify to this:

```
var sizeQualList = calSizeQualityList(metadata);
resolve(parsedImgArr, sizeQualList);  
```

**Pull request merged**